Fix #11273: 32 bits compilation broken due to invalid Stack_ctx_words

fabbing opened the pull request:
https://github.com/ocaml/ocaml/pull/11273 didn't take into account the 32 bits architecture when updating `Stack_ctx_words`.
`Stack_ctx_words` depends on word size and struct padding may bite, details are in the conversation of https://github.com/ocaml/ocaml/pull/11273.

fabbing committed:
Fix Stack_ctx_words computation not architecture depedent

xavierleroy commented on runtime/caml/config.h:
```diff
@@ -13,6 +13,8 @@
 /*                                                                        */
 /**************************************************************************/
 
+#include <limits.h>
```
This breaks the build on Windows mingw (32 and 64 bits), because <limits.h> contains things that ocamlc doesn't understand:
```
File "C:\builds\workspace\precheck\flambda\false\label\ocaml-mingw-32\utils\domainstate.ml", line 1, characters 0-1:
1 | #pragma pack(push,_CRT_PACKING)
```
In the current state of the implementation, you cannot include any standard include in `config.h` because it has to be parsable by OCaml after preprocessing.

xavierleroy reviewed (commented):
Thanks for this new attempt.  Of course I ran a round of CI precheck, and it reveals a new problem on ARM 32 bits and PowerPC 32 bits:
```
/home/touraine/ci/builds/workspace/precheck/flambda/false/label/ocaml-arm-32/runtime/fiber.c:324: undefined reference to `__atomic_fetch_add_8'
```
These platforms don't have 64-bit atomic operations, so they implement `_Atomic int64_t` with a mutex, and then call into library functions that are automatically provided by clang, but require to link with  `-latomic` in GCC.

The Mingw 32 port shows that something else is wrong, see below.

fabbing committed:
Use ARCH_INT64_TYPE instead of int64_t

fabbing committed:
Use __CHAR_BIT__ instead of CHAR_BIT

In order to get a compile time value of CHAR_BIT without including
<limits.h>.
But this is a gcc predefined macro (also supported by clang).

fabbing committed:
configure.ac checks for _Atomic int64_t and links with atomic otherwise

fabbing committed:
Generated configure

xavierleroy commented:
After playing a few rounds of CI golf, I just pushed a reasonable fix directly to trunk, commits 2d8a0cfb53 and 0741d0c017.

fabbing commented:
Oh, I didn't notice your commits while I was fiddling with `configure.ac`.
Glad the CI is green again, thanks for the help!
